examples/cis3-nft-sponsored-txs/src/lib.rs

```diff
@@ -0,0 +1,2228 @@
+/// function of the state. Logs a `Transfer` event and invokes a receive hook
+/// function.
+fn transfer<S: HasStateApi>(
+    token_id: ContractTokenId,
+    amount: ContractTokenAmount,
+    from: Address,
+    to: Receiver,
+    data: AdditionalData,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    logger: &mut impl HasLogger,
+) -> ContractResult<()> {
+    let (state, builder) = host.state_and_builder();
+
+    let to_address = to.address();
+    // Update the contract state
+    state.transfer(&token_id, amount, &from, &to_address, builder)?;
+
+    // Log transfer event
+    logger.log(&Cis2Event::Transfer(TransferEvent {
+        token_id,
+        amount,
+        from,
+        to: to_address,
+    }))?;
+
+    // If the receiver is a contract: invoke the receive hook function.
+    if let Receiver::Contract(address, function) = to {
+        let parameter = OnReceivingCis2Params {
+            token_id,
+            amount,
+            from,
+            data,
+        };
+        host.invoke_contract(&address, &parameter, function.as_entrypoint_name(), Amount::zero())?;
+    }
+
+    Ok(())
+}
+
+/// Execute a list of token transfers, in the order of the list.
+///
+/// Logs a `Transfer` event and invokes a receive hook function for every
+/// transfer in the list.
+///
+/// It rejects if:
+/// - It fails to parse the parameter.
+/// - Any of the transfers fail to be executed, which could be if:
+///     - The `token_id` does not exist.
+///     - The sender is not the owner of the token, or an operator for this
+///       specific `token_id` and `from` address.
+///     - The token is not owned by the `from`.
+/// - Fails to log event.
+/// - Any of the receive hook function calls rejects.
+#[receive(
+    contract = "cis3_nft",
+    name = "transfer",
+    parameter = "TransferParameter",
+    error = "ContractError",
+    enable_logger,
+    mutable
+)]
+fn contract_transfer<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+    logger: &mut impl HasLogger,
+) -> ContractResult<()> {
+    // Parse the parameter.
+    let TransferParams(transfers): TransferParameter = ctx.parameter_cursor().get()?;
+    // Get the sender who invoked this contract function.
+    let sender = ctx.sender();
+
+    for Transfer {
+        token_id,
+        amount,
+        from,
+        to,
+        data,
+    } in transfers
+    {
+        // Authenticate the sender for this transfer
+        ensure!(
+            from == sender || host.state().is_operator(&sender, &from),
+            ContractError::Unauthorized
+        );
+
+        transfer(token_id, amount, from, to, data, host, logger)?;
+    }
+    Ok(())
+}
+
+/// The parameter type for the contract function `permitTransfer`.
+#[derive(Debug, Serialize, SchemaType)]
+pub struct PermitTransferParams(#[concordium(size_length = 2)] pub Vec<PermitTransferParam>);
```

Comment:
I would limit the entry point to only take one signed message instead of a list.

Also, the signed message should contain a list of transfers, like in CIS-2.
Currently, you cannot sign multiple transfer in one message, and signing multiple messages allows the account sponsoring the transaction to split it up, which might not be ideal.